### DISCUSSION TITLE
Combine Easy Effects source (processed mic input) with another audio source? (OBS Monitor)

### DISCUSSION THREAD

**Discussion by polarathene** [wwmm/easyeffects#1957] [Q&A] [answered]
I use OBS for managing a virtual camera, and would like have the ability to change to a scene in OBS that plays a video with audio.

The video appears on the virtual camera, but as apps like Discord and Zoom are using my default mic device for the audio, I need to figure out how to combine the OBS monitor output (or another audio output) into a virtual source that can be mixed with my microphones input (after processing ideally).

I don't want most of the processing from Easy Effects on my mic affecting the audio from this additional stream, so I assume I should create a virtual source that can take both the resulting Easy Effects Source device (or default device?) and the application stream (OBS Monitor), and then I should have a new input device I can have Discord + Zoom use?

I have seen the maintainer of Easy Effects advise not to use Easy Effects Source, but prefer the default device? But as Easy Effects has been explained as doing some "hijacking" here with that default device, I'm not sure if that complicates things here?

---

Resources:

- https://superuser.com/questions/1621243/combining-selected-applications-sound-with-voice-into-virtual-microphone-linux
- https://beko.famkos.net/2022/04/18/virtual-sinks-and-mic-with-obs-and-pipewire-on-linux-pc/
- https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Virtual-Devices#coupled-streams
- https://github.com/wwmm/easyeffects/discussions/1288

Looks like I can use some commands detailed in those resources to achieve this, or use a PatchBay UI app.

Just seeking guidance on what input source I should be connecting from the Easy Effects side into this new virtual source / mic device?

**Comment by wwmm** [comment, marked as answer]
>I have seen the maintainer of Easy Effects advise not to use Easy Effects Source, but prefer the default device? 

That is for the case where people open apps like Pavucontrol or Gnome Settings to configure the `system default device`. Selecting a device in OBS menu is fine. What should be avoided is setting EasyEffects virtual devices as the `whole system` default devices. The main reason for this is that when EasyEffects is initializing it asks PipeWire about the system default devices. It you set our virtual devices as default EasyEffects may not know which device it should use in some situations.

>Just seeking guidance on what input source I should be connecting from the Easy Effects side into this new virtual source / mic device?

If you just want to mix the processed mic signal to something else I think that the easiest thing to do would be loading a `null sink` and linking the output of EasyEffects `virtual source` to the null sink input. Do the same to the other signal source you want to mix.

**Reply by polarathene** [reply]
<details>
<summary>List of devices via pw-link to choose from</summary>

**`pw-link -o`:**

```
Midi-Bridge:Midi Through:(capture_0) Midi Through Port-0
v4l2_input.pci-0000_00_14.0-usb-0_8_1.0:out_0
libcamera_input.__SB_.PCI0.XHC_.RHUB.SS08-8_1.0-0fd9_0078:out_0

alsa_output.usb-Deity_Deity_VO-7U_206833845856-00.analog-stereo:monitor_FL
alsa_output.usb-Deity_Deity_VO-7U_206833845856-00.analog-stereo:monitor_FR
alsa_input.usb-Deity_Deity_VO-7U_206833845856-00.analog-stereo:capture_FL
alsa_input.usb-Deity_Deity_VO-7U_206833845856-00.analog-stereo:capture_FR

alsa_output.pci-0000_00_1f.3.iec958-stereo:monitor_FL
alsa_output.pci-0000_00_1f.3.iec958-stereo:monitor_FR

easyeffects_sink:monitor_FL
easyeffects_sink:monitor_FR
easyeffects_source:capture_FL
easyeffects_source:capture_FR

ee_soe_output_level:output_FL
ee_soe_output_level:output_FR
ee_soe_spectrum:output_FL
ee_soe_spectrum:output_FR
ee_soe_stereo_tools:output_FL
ee_soe_stereo_tools:output_FR
ee_soe_rnnoise:output_FL
ee_soe_rnnoise:output_FR
ee_soe_limiter:output_FL
ee_soe_limiter:output_FR
ee_sie_output_level:output_FL
ee_sie_output_level:output_FR
ee_sie_spectrum:output_FL
ee_sie_spectrum:output_FR
ee_test_signals:output_FL
ee_test_signals:output_FR
ee_sie_rnnoise:output_FL
ee_sie_rnnoise:output_FR
ee_sie_speex:output_FL
ee_sie_speex:output_FR
ee_sie_gate:output_FL
ee_sie_gate:output_FR
ee_sie_autogain:output_FL
ee_sie_autogain:output_FR
ee_sie_autogain:output_FL
ee_sie_autogain:output_FR
ee_sie_equalizer:output_FL
ee_sie_equalizer:output_FR
ee_sie_limiter:output_FL
ee_sie_limiter:output_FR

OBS:monitor_FL
OBS:monitor_FR
OBS:monitor_FL
OBS:monitor_FR
OBS:monitor_FL
OBS:monitor_FR
OBS-Monitor:output_FL
OBS-Monitor:output_FR
```

That is the bulk of the output from the command.

I have seen in Helvum that there is multiple OBS nodes as this output list above also shows, similar for some Input device effects in Easy Effects UI where I have more than one instance of an effect such as Autogain.

> output of EasyEffects `virtual source` to the null sink input.

This was what I was seeking clarity on. The default input that Easy Effects uses for the mic is the Deity VO-7U device.

**`pw-link -i`:**

```
pw-link -i

Midi-Bridge:Midi Through:(playback_0) Midi Through Port-0
alsa_output.usb-Deity_Deity_VO-7U_206833845856-00.analog-stereo:playback_FL
alsa_output.usb-Deity_Deity_VO-7U_206833845856-00.analog-stereo:playback_FR

alsa_output.pci-0000_00_1f.3.iec958-stereo:playback_FL
alsa_output.pci-0000_00_1f.3.iec958-stereo:playback_FR

easyeffects_sink:playback_FL
easyeffects_sink:playback_FR
easyeffects_source:input_FL
easyeffects_source:input_FR
```

So there is `monitor` / `capture` for the outputs, and `playback` / `input` for inputs on each device. Which kind and device is meant to be targeted here from Easy Effects `virtual source`?

~~You had mentioned that Easy Effects hijacks my selected (_or default_) input device (_the Deity mic_), and apps choose that instead of the "Easy Effects Source" input. Should I be connecting that device, or is this a situation where it is ok to connect to "Easy Effects Source"?~~ (See EDIT below)

</details>

My best guess is `easyeffects_source:capture_*` (FL + FR) as it mentions the `source` and presumably that is processed mic signal?

---

> What should be avoided is setting EasyEffects virtual devices as the `whole system` default devices. The main reason for this is that when EasyEffects is initializing it asks PipeWire about the system default devices.

**EDIT:** Reading over that quote again, I think that `easyeffects_source:capture_*` (FL + FR) is probably what you meant by `virtual source`, as you have clarified that apps can select "Easy Effects Source" for input, it's only system default input should not be set to "Easy Effects Source" :+1: 

---

## Solution

I have created two `null-sink` audio devices with fairly generic name prefixed with `x-` while trying to understand this (_easier to grep / search for among all the other devices_). I am not sure if this is the right way to accomplish this, but it seems to work.

I referenced the instructions [here](https://superuser.com/questions/1675877/how-to-create-a-new-pipewire-virtual-device-that-to-combines-an-real-input-and-o/1687100#1687100) with some earlier resources linked before.

The [Beko OBS PipeWire guide](https://beko.famkos.net/2022/04/18/virtual-sinks-and-mic-with-obs-and-pipewire-on-linux-pc/) makes a bit more sense now, I originally was confused by the command snippets as it looked like there was a familiar `pactl` command, followed by some other lines that didn't look like commands... but that was just formatting and I realize now belongs to the previous `pactl` commands :sweat_smile: (_thanks for the guide @bekopharm ! It helped understand the steps needed_)

```bash
# Create a virtual sink to combine multiple audio streams, and a virtual source (mic) for apps to receive that virtual sinks audio from:
pactl load-module module-null-sink media.class=Audio/Sink sink_name=x-virtual-sink-combined
pactl load-module module-null-sink media.class=Audio/Source/Virtual sink_name=x-virtual-mic

# Connect the Easy Effects virtual source (mic after effects processing) `capture` channels to the virtual sink `playback` channels:
pw-link easyeffects_source:capture_FL x-virtual-sink-combined:playback_FL
pw-link easyeffects_source:capture_FR x-virtual-sink-combined:playback_FR

# Connect another audio stream from an app like the `OBS-Monitor`:
pw-link OBS-Monitor:output_FL x-virtual-sink-combined:playback_FL
pw-link OBS-Monitor:output_FR x-virtual-sink-combined:playback_FR

# Connect the sinks `monitor` channels to the virtual mic `input` channels:
pw-link x-virtual-sink-combined:monitor_FL x-virtual-mic:input_FL
pw-link x-virtual-sink-combined:monitor_FR x-virtual-mic:input_FR
```

This resulted in a node graph like this in QPWGraph:

![Screenshot_20221112_105757](https://user-images.githubusercontent.com/5098581/201439342-fd87922f-b433-4284-a2a6-a09b1b61df5e.png)

But when I tried to use this new virtual mic in Discord:

![Screenshot_20221112_112725](https://user-images.githubusercontent.com/5098581/201439417-7a51a329-c270-4fa4-9287-e2bd96e8dc3b.png)

There was no audio being detected... If I unmuted my mic (physical button) and disabled the Noise Reduction + Gate effects, it would pick up on the mic, but the other channel connected to the device would still not seem to have any effect. Looked at QPWGraph to troubleshoot and:

![Screenshot_20221112_113314](https://user-images.githubusercontent.com/5098581/201439898-257d17bc-03a6-4290-b26e-920bcbac5bff.png)

Despite choosing the explicit virtual mic device created, the "WEBRTC VoiceEngine" for Discords `input` channels are being connected to "Easy Effects Source" `capture` channels, not "x-virtual-mic" `capture` channels like expected..

**Solution:** Manually disconnecting `capture` channels from "Easy Effects Source" and replacing them with "x-virtual-mic" `capture` channels instead, now Discord correctly gets the expected audio mix:

![Screenshot_20221112_113723](https://user-images.githubusercontent.com/5098581/201440208-1288f6a3-e76a-48f4-b211-0df869ec8d1f.png)

**Reply by polarathene** [reply]
## Problem

Is Easy Effects doing some hijacking here when I select an explicit input device in Discord?

What would cause Discord to ignore my explicit choice and connect to the "Easy Effects Source" `capture` channels instead?

**Reply by polarathene** [reply]
> Is Easy Effects doing some hijacking here when I select an explicit input device in Discord?

**Answer:** Yes. Explained by @wwmm [here](https://github.com/wwmm/easyeffects/discussions/1288#discussioncomment-1813123).

![Screenshot_20221112_115909](https://user-images.githubusercontent.com/5098581/201441940-6dcbd85f-d459-4307-bd4d-edceb649416d.png)

- "WEBRTC VoiceEngine" would appear here when using Discord (_at least when enabling mic, such as the settings page testing_).
- The "Enable" checkbox was not ticked after I manually rewired the connection in QPWGraph, but the next time it was switched Easy Effects took over again.
- Clicking "Exclude" checkbox adds it to the "Excluded Apps" in the bottom right.

![Screenshot_20221112_120443](https://user-images.githubusercontent.com/5098581/201442345-17ab1e0f-dcca-4a52-99a9-15af8ca6a7c4.png)

That will prevent "Default" from working as well, as it excludes the "App" not a specific device. Instead an app would have to explicitly choose "Easy Effects Source" as the input device, which still works even when the "App" is excluded, the "Enable" checkbox becomes forced and the connection is made:

![Screenshot_20221112_120318](https://user-images.githubusercontent.com/5098581/201442253-f57c93c2-7ef4-4f3e-a331-7b5a0e40ca8f.png)

So that's good :+1:

---

Just that warning in Easy Effects "PipeWire" tab that gave the impression that we should avoid selecting this input :sweat_smile:

![Screenshot_20221112_120904](https://user-images.githubusercontent.com/5098581/201442637-82692662-4ea6-4d75-9b48-fa72263c4fa6.png)

But we've since cleared that up, that "external applications" (as the examples try to make apparent) is only about applications that set the default audio device on a system, not apps that choose a mic or speaker to use for themselves. And that this message may get further revised in a future release :+1:

**Reply by wwmm** [reply]
>Reading over that quote again, I think that easyeffects_source:capture_* (FL + FR) is probably what you meant by virtual source, as you have clarified that apps can select "Easy Effects Source" for input, it's only system default input should not be set to "Easy Effects Source"

Yes.

>Is Easy Effects doing some hijacking here when I select an explicit input device in Discord?

>What would cause Discord to ignore my explicit choice and connect to the "Easy Effects Source" capture channels instead?

Is `Process All inputs` enabled in EasyEffects preferences window? If yes it will force Discord to be moved to EasyEffects virtual source. THere are two solutions for this. Either do not process all inputs. What can be annoying. Or add Discord to the `Input section blocklist`. This will tell EasyEffects to not try to move Discord streams that are recording audio.

>Clicking "Exclude" checkbox adds it to the "Excluded Apps" in the bottom right.

Yes. This will add Discord to the blocklist.

**Reply by polarathene** [reply]
> Is `Process All inputs` enabled in EasyEffects preferences window?

Yes that was enabled. I'll keep it enabled for now.

- So this is the feature that is "hijacking" my audio devices when other apps select them and rerouting to Easy Effects source/sink nodes?
- If I didn't use that, then each app would have to explicitly opt-in by selecting an "Easy Effects ..." device for input/output?

> Yes. This will add Discord to the blocklist.

As it's called "WEBRTC Voice Engine" which seems rather generic - is it specifically Discord? Or anything else that might use this for voice chat that such as a browser or desktop apps like those built off Electron (Chromium)?

I assume there is some metadata somewhere that might differentiate instances of nodes like that. In Helvum and QPWGraph GUI apps I have found this a bit confusing to discern. Especially with "Easy Effects Source" that has been getting all these extra "PulseAudio Volume Control" nodes:


![Screenshot_20221112_141040](https://user-images.githubusercontent.com/5098581/201449424-35f3bc38-a37f-4589-811b-90e710bef287.png)

**Reply by wwmm** [reply]
>As it's called "WEBRTC Voice Engine" which seems rather generic - is it specifically Discord? Or anything else that might use this for voice chat that such as a browser or desktop apps like those built off Electron (Chromium)?

Oh... I forgot that Discord does not set a proper name to its node. Unfortunately anything with this name will be blocklisted.

>Especially with "Easy Effects Source" that has been getting all these extra "PulseAudio Volume Control" nodes:

These seem to come from `Pavucontrol` level meter streams. Closing Pavucontrol should make them go away.

**Reply by polarathene** [reply]
> Discord does not set a proper name to its node. Unfortunately anything with this name will be blocklisted.

Good to know, I'll keep that in mind thanks!

> These seem to come from Pavucontrol level meter streams. Closing Pavucontrol should make them go away.

That was it! :sweat_smile: 

Thank you for being so quick to respond! :)

**Reply by bekopharm** [reply]
> (_thanks for the guide @bekopharm ! It helped understand the steps needed_)

You're welcome :) Always happy when my scribbles helped someone. Totally worth it :grin:

**Reply by bekopharm** [reply]
> Oh... I forgot that Discord does not set a proper name to its node. Unfortunately anything with this name will be blocklisted.

Pretty sure that's a browser thing. I noticed the same in Jitsi meetings in Firefox :shrug: 
Would be great if they'd set it to the tab title or something :thinking:

**Comment by polarathene** [comment]
Sharing extra information here in case it helps anyone else :)

## Gotcha - Multiple OBS monitors (one per source) each require routing

I have noticed with OBS that each source added (_media source in my case_) that adds an entry into the "Audio Mixer" adds it's own monitor node (_if you toggle on monitoring for that audio stream_).

The "output" option in OBS for these audio streams presumably only adds an output node when streaming / recording - which I am doing neither.

I use the virtual camera feature and wanted the ability to switch to scenes with audio mixed in for video calls, so only the "monitor" nodes will be available (when enabled).

---

### OBS setting - Default Monitor device to route each OBS monitor into

To avoid linking each one via `pw-link` commands like above, or manually routing the channel connections each time with a patchbay GUI app, OBS settings has a default monitor device to connect those OBS monitor nodes to (_as Beko's guide shows in screenshots_).

This way the `x-virtual-sink-combined` node can be chosen to automatically connect to that. **_Easy Effects Output Players_ view must exclude "OBS-Monitor" for this to work**, otherwise it will keep getting routed into Easy Effects Sink.

![Screenshot_20221112_180455](https://user-images.githubusercontent.com/5098581/201458178-a31305eb-379a-419a-bf63-49521520dcdd.png)

![Screenshot_20221112_180832](https://user-images.githubusercontent.com/5098581/201458292-6437d99c-20d1-4fb2-9cc4-c6a1e4974d1f.png)

![Screenshot_20221112_181519](https://user-images.githubusercontent.com/5098581/201458460-f2ce389f-23e3-44fb-bdb3-3a207f2ac03a.png)

![Screenshot_20221112_181630](https://user-images.githubusercontent.com/5098581/201458534-a7bfe7ef-12b5-4fe7-a015-43af00922c1b.png)

---

## OBS Limiter filter

Another issue I noticed is depending on the media you might need to account for different volume levels. One could manually adjust the media's volume sliders in OBS, or apply a common Limiter Filter as a way around that.

Beko's guide also shows having the virtual sinks in the OBS Audio Mixer settings, "Audio Output Capture (PulseAudio)" source can be added to choose one of those virtual sinks, then each scene can use that source (_copy / paste within OBS to each scene shares a reference to the same node in this case_). That source can then be the only one that has the Limiter Filter added, which might be more optimal (_and easier to manage than adding the same filter to each source_).

![Screenshot_20221112_183618](https://user-images.githubusercontent.com/5098581/201459112-0bd003dc-dc35-45f9-8de9-8bd7f75aa767.png)

This source connects a new node in PipeWire to your virtual sink, which you can then route into the `x-virtual-mic` - and if you want to monitor it, into the "Easy Effects Sink".

The "Easy Effects Source" should probably go directly into `x-virtual-mic` at that point, making the `x-virtual-sink-combined` specifically for the OBS streams, unless you want all other merged audio streams to go through that OBS limiter too.

![Screenshot_20221112_190225-1](https://user-images.githubusercontent.com/5098581/201460403-048f4dee-572d-42b2-9e41-88dc4b2052c6.png)

**NOTE:** I am not sure if it'll make much difference vs the PulseAudio Output Capture source, but there is a [third-party option for PipeWire equivalent (`obs-pipewire-audio-capture`)](https://github.com/dimtpap/obs-pipewire-audio-capture), which will eventually be upstreamed into a future OBS release.

---

### Easy Effects Sink / Output for applying a Limiter instead of via OBS

Instead of the OBS Limiter Filter, if including all system / desktop audio is ok - presumably you could keep the default "Easy Effects Sink" as the OBS Monitor, and instead route the processed signal from Easy Effects "Output" into the `x-virtual-mic` device instead. Allowing to use the Limiter effect there plus many other effects.

It's unclear which node is the equivalent of "Easy Effects Source" for that _processed Output_ however. It seems like it might be the `output_level` node when looking in QPWGraph at the chain of nodes from "Easy Effects Sink" monitor channels.

It's probably better to just stick with the OBS limiter as Easy Effects does not support configuring effects per Output / Input devices, only a single device.

**Reply by polarathene** [reply]
Not quite solved yet, the commands that created the virtual nodes turns out to be temporary, and they are lost during a restart. You must do that all over again each restart :thinking: 

Next step I guess is to look into configuration files to handle that, will share when I have more time to investigate :)
